L1-regularization, regularization path, plot, and predict method

jolars opened the pull request:
I apologize in advance for the somewhat messy commit history for this pull request. I hope you are able to follow along okay.

# L1-regularization
is now supported and in theory the elasticnet penalty should too, although it does not seem to work quite as expected.

# Regularization path
`sgdnet()` now computes and fits a lambda path as `glmnet()` does, using warm starts. This meant I had to split up a lot of the code in the former `SagaSolver()`, which is now spread into `SetupSaga()` and `Saga()` (which holds the actual algorithm).

# Plotting
I have set up a plotting method for the `'sgdnet'` object, which is very similar to what glmnet does but based on lattice instead of base graphics.

# Predict method
There is now a `predict.sgdnet()` method, which is basically a copy of `predict.glmnet()`. I have updated COPYRIGHTS and prepended a copyright notice regarding this.

# Example
Here follows a brief example of the package as it stands, and also where it currently fails.

```
library(sgdnet)
library(glmnet)

x <- as.matrix(iris[, 2:4])
y <- iris[, 1]

glmnetfit1 <- glmnet::glmnet(x, y, alpha = 1)
sgdnetfit1 <- sgdnet::sgdnet(x, y, alpha = 1)

# similar results
plot(glmnetfit1)
```
![image](https://user-images.githubusercontent.com/13087841/40629235-1d0104da-62ca-11e8-9ece-d3fa60b21c4e.png)
```
plot(sgdnetfit1)
```
![image](https://user-images.githubusercontent.com/13087841/40629240-25b78f40-62ca-11e8-9c8c-2c57247d6306.png)

```
glmnetfit2 <- update(glmnetfit1, alpha = 0)
sgdnetfit2 <- update(sgdnetfit1, alpha = 0)

# seems okay too, although we have no stopping criteria as glmnet does yet
plot(glmnetfit2)
```
![image](https://user-images.githubusercontent.com/13087841/40629268-4f4ed43a-62ca-11e8-8109-81c4e144b564.png)
```
plot(sgdnetfit2)
```
![image](https://user-images.githubusercontent.com/13087841/40629279-60581782-62ca-11e8-97c3-a0a64360a276.png)
```
glmnetfit3 <- update(glmnetfit1, alpha = 0.5)
sgdnetfit3 <- update(sgdnetfit1, alpha = 0.5)

# this is wrong though
plot(glmnetfit3)
```
![image](https://user-images.githubusercontent.com/13087841/40629298-78c6ed20-62ca-11e8-9beb-1dcf8007ef67.png)
```
plot(sgdnetfit3)
```
![image](https://user-images.githubusercontent.com/13087841/40629302-854e2c48-62ca-11e8-9bdb-c6b93f1e6b7f.png)

```

jolars committed:
Fit models to a path of regularization values using warm starts for subsequent fits and more

This commit enables fitting a ful regularization path of lambda values. We now store weights, intercept, and gradients between separate fits to speed up convergence.

This commit is unfortunately somehwat muddled with additional changes, which include

* Now using smart pointers
* Setting up factories to create Family and Prox objects. The Family class now replaces the former Objective class and provides the additional functionality of setting the number of classes, as well as computing step size.
* Changing name of the main function to SgdnetCpp, which is only there to provide branching of sparse and dense feature implementations, and then passes along all other parameters to SetupModel, which sets up the specifics of the model. Finally, Saga() is called repeatedly for each lambda value to fit the model to the entire path.

jolars committed:
Add a plotting method for sgdnet objects

This plotting method mimics plot.glmnet() but uses lattice graphics instead of base graphics to make it easier to deal with the output since it is now a grid/trellis object and makes it much easier to modify the arguments of the underlying function since we are calling a single function behind the scenes (xyplot()).

We also add a simple test for the gaussian regression to check that we retrieve a trellis object. We will also need to use better plot tests where the actual plotting mechanism is called eventually, but for now I think this will do.

jolars committed:
Redesign coef.sgdnet work with the new structure of the sgdnet objects.

jolars committed:
Rearrange output to mimic that of glmnet

jolars committed:
Prepend copyright information

jolars committed:
Shed off some empty lines

jolars committed:
Update documentation

jolars committed:
Change return name for debug_info to diagnostics

jolars committed:
Add @noRd and internal keyword

jolars committed:
Fix error in example

Features and response got mixed up.

jolars committed:
Use Matrix::Matrix() instead of as() to convert matrixes from dense to sparse

jolars committed:
Correct scaling and setup regularization path function

jolars committed:
Preprocess features separate and response for each family.

jolars committed:
Plot reference line and prettier x axis label

jolars committed:
Use fixed size steps vector and correct erroneous evaluation on prox

In addition, there are a few formatting changes that have no real consequence.

jolars committed:
Setup arguments for automatic computation of lambda path on the R side

jolars committed:
Drop extraneous brackets

jolars committed:
Add tests comparing glmnet regularization path with sgdnet

jolars committed:
Suppress plotting output from tests using dont_plot() and check xvar argument in plotting

jolars committed:
Redesign and expands tests around lambda_max

jolars committed:
Drop empty line

jolars committed:
Add predict methods for the gaussian family

Add predict.sgdnet() by borrowing significantly from predict.glmnet().

We also
* add tests for this,
* store the call in sgdnet() so that we can use update() on the resulting object,
* rewrite coef.sgdnet() to call predict.sgdnet() with type = "coefficients", and
* update COPYRIGHTS so that we provide credit where it is due.

jolars committed:
Correct broken link to package documentation

jolars committed:
Rename contexts

jolars committed:
use dash instead of underscore in file name

jolars committed:
Offer a better example for predict.sgdnet

jolars committed:
Use dash instead of underscore

jolars committed:
Extend documentation regarding computation of regularization path and document nlambda and lambda.min.ratio

jolars committed:
Convert data.frames to matrices before predicting. Also provide explicit namespaces.

jolars committed:
Tidy up example

michaelweylandt commented on R/sgdnet.R:
```diff
@@ -16,10 +16,27 @@
+#' log-spaced elements
+#' from \eqn{\log_e{\lambda_{\mathrm{max}}}}{log(lambda_max)} to
+#' \eqn{\log_e{\lambda_{\mathrm{max}}} \times \mathtt{lambda.min.ratio}}{
+#'      log(lambda_max)*lambda.min.ratio},
```
This doesn't seem right - isn't it just log-spaced elements from `lambda.min.ratio * lambda_max` to `lambda_max`? I don't think you actually need the `log` parts here (unless I'm missing something)

michaelweylandt commented on R/sgdnet.R:
```diff
@@ -69,55 +91,92 @@ sgdnet.default <- function(x,
+    a0 <- t(a0)
+    rownames(a0) <- response_names
+    colnames(a0) <- path_names
+  }
```
What exactly is happening here? It looks like you're transposing the intercept vectors for the multi-coefficient families, but I'm not sure why you'd do that 

I don't think you need to match the internal data structures of `glmnet` exactly - only the API.

jolars replied:
Okay, yeah, that was the purpose of this. I guess it would be alright if I also change the return type of the coefficients? I am personally not happy that the coefficients are sometimes a list (mgaussian and multinomial) and sometimes a matrix (as well as the direction of that matrix, having the features as rows) and would much rather always return a 3-dimensional array (cube).

This is really part of a greater discussion regarding the package and I'll post an issue so that we can discuss if and when we would like to depart from glmnet.

michaelweylandt replied:
I agree the `glmnet` design is weird here, but I understand what they're going for - sparse 3d array libraries are rare (`R` recently got them through the `slam` package and Armadillo doesn't have a sparse cube type). 

I lean towards sticking to `glmnet`-formatted stuff for the output of accessor functions (`coef`) while saying that we have the freedom to define internal structures however we want, but happy to discuss in a different issue.

michaelweylandt commented on src/families.h:
```diff
@@ -1,10 +1,110 @@
-  enum Family {
-    GAUSSIAN = 0
+
+class Family {
```
I like this design a lot.

michaelweylandt commented on src/families.h:
```diff
@@ -1,10 +1,110 @@
+      y_center = arma::mean(y);
+      //y_scale  = arma::sqrt(arma::var(y, 1));
+
+      for (arma::uword i = 0; i < y.n_cols; ++i) {
```
Why is `y` a matrix here as opposed to a vector?

jolars replied:
Because it will be a matrix for the multivariate case. We could template/overload or drop the additional dimension for the other cases using arma::conv_to<arma::vec>::from(x) but I am not sure it provides any meaningful performance boost.

michaelweylandt replied:
Ok -- looks good. I was more confused as to why that one was a matrix, but the other methods were vectors. I see you've changed it elsewhere.

michaelweylandt commented on tests/testthat/test-predictions.R:
```diff
@@ -0,0 +1,38 @@
+
+  # check linear interpolation
+  pred_new <- predict(fit, x, s = 0.04)
+  expect_is(pred_new, "matrix")
```
Can we check that these tests are statistically right in addition to returning the right type of objects?

jolars replied:
Could you please specify what you are after? I've added a test for the linear predictors but checking the weights seems just like repeating what we are already doing in [test-gaussian.R](https://github.com/jolars/sgdnet/blob/master/tests/testthat/test-gaussian.R), no?

michaelweylandt replied:
For example, here you test that the interpolation returns a matrix, but you don't actually check that it's statistically correct. You should also have a test along the lines of  `expect_equal(pred_new, 0.5 * predict(fit, x, s=0.03) + 0.5 * predict(fit, x, s=0.05))` [that's not actually correct, but you get the idea.] I don't think this is in the Gaussian tests yet, but I might be missing it. 

Similarly, to check pred_nonzero and pred_coefficients, I'd at least check something like `expect_equal(pred_nonzero, pred_coefficients != 0)`.

michaelweylandt reviewed (commented):
There's a lot here, so I'll come back and make another pass, but overall I think this looks good.

jolars committed:
Correct erroneous specification of regularization path

jolars committed:
Correct extraction of coefficients when there is only one feature vector

jolars committed:
Standardize the outcome for the gaussian model as well.

jolars committed:
Correct loss computation

It was missing the l1-penalty and used the wrong l2-penalty.

jolars committed:
Add a test for the debug option

jolars committed:
Expand with a few more tests on using predict.sgdnet()

jolars committed:
Fix reorganization of coefficients to work for all cases

jolars committed:
Don't condition on feature type when rescaling

The center and scale vectors are set appropriately so that nothing bad happens anyway.

jolars committed:
Remove now-useless conversion.

jolars committed:
Add a few more tests for predict.sgdnet

jolars committed:
Change = to <-

jolars committed:
Remove y argument from generic to possibly enable other methods dispatching only on x in the future.

jolars committed:
Allow valgrind build to fail (for now) since it is timing out

jolars commented:
@michaelweylandt and @tdhock, do you have any other comments regarding this pull request? Otherwise I suggest merging so we can keep on working on the upcoming features? I don't think there are any glaring issues here right now except for the inconsistencies compared to glmnet with the elastic net penalty.

michaelweylandt commented:
I'm generally a bit hesitant to merge PRs with known bugs, but since the elastic net issue is addressed in #9 I'm good with merging this one.

jolars commented:
I agree; it's bad form, but let's do it anyway because it is practical here.
